packages/common/src/grpc-operation.js

```diff
@@ -0,0 +1,212 @@
+        method: 'deleteOperation'
+      },
+      reqOpts: {
+        name: name
+      }
+    },
+
+    /**
+     * Checks to see if an operation exists.
+     */
+    exists: true,
+
+    /**
+     * Retrieves the operation.
+     */
+    get: true,
+
+    /**
+     * Retrieves metadata for the operation.
+     */
+    getMetadata: {
+      protoOpts: {
+        service: 'Operations',
+        method: 'getOperation'
+      },
+      reqOpts: {
+        name: name
+      }
+    }
+  };
+
+  var config = {
+    parent: parent,
+    id: name,
+    methods: methods
+  };
+
+  GrpcServiceObject.call(this, config);
+  events.EventEmitter.call(this);
+
+  this.completeListeners = 0;
+  this.hasActiveListeners = false;
+
+  this.listenForEvents_();
+}
+
+modelo.inherits(GrpcOperation, GrpcServiceObject, events.EventEmitter);
+
+/**
+ * Status text used to describe the current state of the operation.
+ *
+ * @private
+ */
+GrpcOperation.STATUS = {
+  running: 'RUNNING',
+  done: 'DONE'
+};
+
+/**
+ * Cancel the operation.
+ *
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this
+ *     request.
+ * @param {object} callback.apiResponse - The full API response.
+ */
+GrpcOperation.prototype.cancel = function(callback) {
+  var protoOpts = {
+    service: 'Operations',
+    method: 'cancelOperation'
+  };
+
+  var reqOpts = {
+    name: this.id
+  };
+
+  this.request(protoOpts, reqOpts, callback || util.noop);
+};
+
+/**
+ * Begin listening for events on the operation. This method keeps track of how
+ * many "complete" listeners are registered and removed, making sure polling is
+ * handled automatically.
+ *
+ * As long as there is one active "complete" listener, the connection is open.
+ * When there are no more listeners, the polling stops.
+ *
+ * @private
+ */
+GrpcOperation.prototype.listenForEvents_ = function() {
+  var self = this;
+
+  this.on('newListener', function(event) {
+    if (event === 'complete') {
+      self.completeListeners++;
+
+      if (!self.hasActiveListeners) {
+        self.hasActiveListeners = true;
+        self.startPolling_();
+      }
+    }
+  });
+
+  this.on('removeListener', function(event) {
+    if (event === 'complete' && --self.completeListeners === 0) {
+      self.hasActiveListeners = false;
+    }
+  });
+};
+
+/**
+ * Poll `getMetadata` to check the operation's status. This runs a loop to ping
+ * the API on an interval.
+ *
+ * Note: This method is automatically called once a "complete" event handler is
+ * registered on the operation.
+ *
+ * @private
+ */
+GrpcOperation.prototype.startPolling_ = function() {
+  var self = this;
+
+  if (!this.hasActiveListeners) {
+    return;
+  }
+
+  this.getMetadata(function(err, resp) {
+    if (err || resp.error) {
+      self.emit('error', err || GrpcService.decorateStatus_(resp.result));
+      return;
+    }
+
+    if (!resp.done && self.status !== GrpcOperation.STATUS.running) {
```

Comment:
Here is what I ended up with after I took out the `status` stuff: https://github.com/stephenplusplus/gcloud-node/blob/4660c1182f903416c19e338884f5a23a217a7bda/packages/common/src/grpc-operation.js